File: 2020/php/src/Aoc/Days/Day9.php
<?php


namespace Ppx17\Aoc2020\Aoc\Days;


class Day9 extends AbstractDay
{
    private const PREAMBLE = 25;
    private int $nc;
    private array $numbers;
    private int $weakness = 0;

    public function dayNumber(): int
    {
        return 9;
    }

    public function setUp(): void
    {
        $this->numbers = array_map(fn($x) => (int)$x, $this->getInputLines());
        $this->nc = count($this->numbers);
    }

    public function part1(): string
    {
        for ($i = self::PREAMBLE; $i < $this->nc; $i++) {
            $current = $this->numbers[$i];
            if (!$this->hasSum($i, $current)) {
                $this->weakness = $current;
                return $current;
            }
        }
        return 'not found';
    }

    private function hasSum(int $index, int $sum): bool
    {
        for ($a = $index - 1; $a >= $index - self::PREAMBLE - 1; $a--) {
            for ($b = $a - 1; $b >= $index - self::PREAMBLE; $b--) {
                if ($this->numbers[$a] + $this->numbers[$b] === $sum) {
                    return true;
                }
            }
        }
        return false;
    }

    public function part2(): string
    {
        $list = [];
        for ($i = 0; $i < $this->nc; $i++) {
            $list[] = $this->numbers[$i];

            if (count($list) === 1) continue;

            while (array_sum($list) > $this->weakness && count($list) > 2) {
                array_shift($list);
            }

            if (array_sum($list) === $this->weakness) {
                return min($list) + max($list);
            }
        }
        return 'not found';
    }
}

File: 2020/julia/src/Day9.jl
module Day9
include("Aoc.jl")
using .Aoc

preamble = 25
numbers = parse.(Int64, Aoc.input_lines(9))
count = size(numbers)[1]
weakness = 0

function part1()
    for (idx, nr) in enumerate(numbers[preamble+1:end])
        if ! has_sum(idx + preamble, nr)
            global weakness = nr
            return nr
        end
    end
end

function part2()
    start = 1
    for i in 1:count
        while sum(numbers[start:i]) > weakness && i - start > 2
            start+=1
        end
        
        if sum(numbers[start:i]) == weakness
            return minimum(numbers[start:i]) + maximum(numbers[start:i])
        end
    end
end

function has_sum(index, sum)
    for a = numbers[index-preamble:index], b = numbers[index-preamble+1:index]
        if a + b == sum
            return true
        end
    end
    return false
end

end